### PR TITLE
`lib.fileset.fileFilter`: Predicate attribute for file extension

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

Makes it more convenient to use `lib.fileset.fileFilter` to filter by file extension:
```nix
lib.fileset.fileFilter (file:
  lib.any file.hasExt [ "nix" "hs" ]
) ./.
```

Whereas previously this would've needed
```nix
lib.fileset.fileFilter (file:
  lib.any (ext: lib.hasSuffix ".${ext}" file.name) [ "nix" "hs" ]
) ./.
```

I'd imagine this should be fairly popular, but we need to confirm first.

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

## Things done

- [x] Tests
- [x] Docs

### CODE REVIEW

**Comment by Kranzes** [comment]
Very looking forward for this

**Comment by infinisil** [comment]
Rebased, added tests and made sure the docs are accurate. I'd say let's have this!

**Comment by roberth** [comment]
> ```nix
> lib.fileset.fileFilter (file:
>   lib.elem file.ext [ "nix" "hs" ]
> ) ./.
> ```

What about `tar.gz`?
How about:

```nix
lib.fileset.fileFilter (file:
  file.hasExt "hs" || file.hasExt "nix"
) ./.
```

```nix
lib.fileset.fileFilter (file:
  ! file.hasExt "tar.gz"
) ./.
```

**Comment by infinisil** [comment]
Not bad, here's another idea though:

```nix
lib.fileset.fileFilter (file:
  file.fullExt != "tar.gz"
) ./.
```

and

```nix
lib.fileset.fileFilter (file:
  file.lastExt != "gz"
) ./.
```

I do think there's some benefit to getting the extension value directly, e.g. you could do
```nix
lib.fileset.fileFilter (file:
  if file.ext == null then
    true
  else
    {
      nix = false;
      c = true;
      h = false;
    }.${file.lastExt}
    or (throw "Don't know what to do with file extension ${file.lastExt}")
) ./.
```

Also there wouldn't be a problem to have multiple of these. Their names make them pretty obvious.

**Comment by roberth** [comment]
But then you have to get into what full means. What if you have

```
foo-master.tar.gz
foo-1.2.tar.gz
```

Is that a `foo-1` man page, archived and compressed?

Determining which part of the file name is extension is up to the caller, not up to `fileFilter`. It can not have the domain knowledge, so it must not try.

**Comment by infinisil** [comment]
Fair enough! I guess with `hasExt`, you can even do

```nix
lib.fileset.fileFilter (file:
  lib.any file.hasExt [ "hs" "nix" ]
) ./.
```

which is pretty short

File: lib/fileset/tests.sh [hunk 1]
```diff
@@ -894,6 +894,37 @@ expectEqual \
...
+# Check that .hasExt checks for the file extension
+# The empty extension is the same as a file ending with a .
+tree=(
+    [a]=0
+    [a.]=1
+    [a.b]=0
+    [a.b.]=1
+    [a.b.c]=0
+)
+checkFileset 'fileFilter (file: file.hasExt "") ./.'
```
Comment (roberth) [inline lib/fileset/tests.sh:906]:
```suggestion
checkFileset 'fileFilter (file: file.hasExt "") ./.'

```

File: lib/fileset/tests.sh [hunk 2]
```diff
@@ -894,6 +894,37 @@ expectEqual \
...
+# It can check for the last extension
+tree=(
+    [a]=0
+    [.a]=1
+    [.a.]=0
+    [.b.a]=1
+    [.b.a.]=0
+)
+checkFileset 'fileFilter (file: file.hasExt "a") ./.'
```
Comment (roberth) [inline lib/fileset/tests.sh:915]:
```suggestion
checkFileset 'fileFilter (file: file.hasExt "a") ./.'

```

File: lib/fileset/tests.sh [hunk 3]
```diff
@@ -894,6 +894,37 @@ expectEqual \
...
+    [a.b.c.d]=1
+)
+checkFileset 'fileFilter (file:
+  all file.hasExt [
+    "b.c.d"
+    "c.d"
+    "d"
+  ]
+) ./.'
+
```
Comment (roberth) [inline lib/fileset/tests.sh:927]: Does `.gitignore` have a `.gitignore` extension, or no extension at all? In other words, does the empty basename exist?

**Review by roberth — COMMENTED** [review]
If a tree falls, but no-one's arou

I think one corner case should be addressed in the docs, and then we're good.

**Comment by fricklerhandwerk** [comment]
Is it a file system tree?

File: lib/fileset/tests.sh (hunk 3 above)
Comment (infinisil) [reply lib/fileset/tests.sh:927]: This is covered with the `file.hasExt "a"` test! Should be added to the docs though :)

Comment (infinisil) [reply lib/fileset/tests.sh:927]: I guess we could ask whether that's the right behavior though, alternatively it could error :thinking:

Comment (infinisil) [reply lib/fileset/tests.sh:929]: I think it's fine to leave it as is, I added some docs for that: [(diff)](https://github.com/NixOS/nixpkgs/compare/5aaac6d99bc48db19916e5e8f2760694a55169cf..6816f28c960c523e6a30f2ad4a1cc812251f5ffb)

**Review by roberth — APPROVED** [review]
The right extent of extension support.👌

**Comment by roberth** [comment]
OfBorg is sluggish today.

> Is it a file system tree?

Nice.

**Comment by infinisil** [comment]
> OfBorg is sluggish today.

Only today? :sweat_smile:
